fontir/src/variations.rs

```diff
@@ -113,6 +114,98 @@ impl VariationModel {
     pub fn default_location(&self) -> &NormalizedLocation {
         &self.default
     }
+
+    /// Convert absolute positions at master locations to offsets.
+    ///
+    /// <ul>
+    ///     <li>All keys in points must be known to the variation model.</li>
+    ///     <li>All point vectors must have the same length.</li>
+    ///     <li>A point vector for the default location is required</li>
+    /// </ul>
+    ///
+    /// Note that it is NOT required to provide a point sequence for every location known
+    /// to the variation model.
+    ///
+    /// P is the point type, meant to be absolute position in 1 or 2 dimensional space.
+    /// V is the vector type.
+    ///
+    /// For 2d [kurbo::Point] and [kurbo::Vec2] would be typical choices. For 1d a floating
+    /// point primitive should suffice.
+    ///
+    /// Returns a delta, as the vector type, for every input point. Intended use is to support
+    /// construction of a variation store.
+    ///
+    /// Rust version of <https://github.com/fonttools/fonttools/blob/3b9a73ff8379ab49d3ce35aaaaf04b3a7d9d1655/Lib/fontTools/varLib/models.py#L449-L461>
+    pub fn deltas<P, V>(
+        &self,
+        point_seqs: &HashMap<NormalizedLocation, Vec<P>>,
+    ) -> Result<HashMap<NormalizedLocation, Vec<V>>, DeltaError>
+    where
+        P: Copy + Default + Sub<P, Output = V>,
+        V: Copy + Mul<f64, Output = V> + Sub<V, Output = V>,
```

Comment:
I've been staring at this for twenty minutes, trying to come up with some compelling alternative, and I'm not having a lot of luck.


Just for the exercise of it, one possible solution I could imagine would be to actually use less fancy types here; we could use const generics to make the input and output of this method always be arrays, where we are generic over the length of the array (it would always be either 1 or 2).

(I went and implemented this and it wasn't really an improvement, so... idk)

Reply:
I followed a similar arc. Ultimately I concluded this did express *exactly* the characteristics I want so ... perhaps this is the intrinsic complexity.